### PR TITLE
SStruct interface without UVM

### PR DESCRIPTION
**Pull request by liruipeng**
In this PR, we removed the dependency of UVM when building with CUDA for the SStruct solvers. The work includes

1. Rewriting certain CPU only boxloops (zypre_BoxLoop) with CPU/GPU boxloops (hypre_BoxLoop)
2. Changing SStruct interface calls such as `hypre_SStructUMatrixSetValues`, where the  input `double *values` is no longer assumed to be able to accessed from the host. In other words, it can be a GPU device-only pointer
3. Removing the use of UVM in drivers `sstruct.c` and `maxwell_unscaled.c`, so that `nvprof` shows no GPU page fault. Theoretically, most of (**not all of**) the SStruct regression test runs can work without UVM.
4. Adding `hypre_ParKrylovCAlloc` having a 3rd `HYPRE_MemoryLocation location` argument
5. Implementing `hypre_ParVectorGetValuesDevice`
6. Adding a "memory tracker" to help debugging memory leak or misuse.

### CODE REVIEW

File: src/sstruct_mv/sstruct_matrix.c [hunk 1]
```diff
@@ -1082,33 +1112,80 @@ hypre_SStructUMatrixSetBoxValues( hypre_SStructMatrix *matrix,
...
                hypre_BoxGetSize(int_box, loop_size);
-               /*FIXME: Currently works only for the default boxloop (see GetIndex below) */
-               zypre_BoxLoop2Begin(ndim, loop_size,
-                                   box,  start, stride, mi,
-                                   value_box, start, stride, vi);
-#ifdef HYPRE_USING_OPENMP
-#pragma omp parallel for private(HYPRE_BOX_PRIVATE) HYPRE_SMP_SCHEDULE
+
+#if defined(HYPRE_USING_CUDA) || defined(HYPRE_USING_DEVICE_OPENMP)
+               hypre_assert(ndim <= 3);
```
Comment (rfalgout) [inline src/sstruct_mv/sstruct_matrix.c:1119]: @liruipeng Can we make this at least 4D?  Are there other 3D limitations in the GPU code?

**Comment by rfalgout** [comment]
Hi @liruipeng .  So far, this looks good to me.  Many of the changes don't seem particularly specific to SStruct, such as the Krylov changes.  There were some changes in utilities about memory trackers that I didn't follow either.  The change from zypre to hypre macros is what we want anyway, as long as the macros themselves work (and I assume they do).  My comment about 4D isn't critical for now, but the original boxloops were written to move beyond 3D to potentially support QCD applications (that rewrite was the reason why the temporary 'zypre_' boxloop name was introduced to begin with), so it would be nice to not re-introduce that limitation.  I also understand that it is more difficult to write a performant GPU loop for arbitrary dimensions.

Let me know when you want me to approve.  As always, nice work!  Thanks!

File: src/sstruct_mv/sstruct_matrix.c (hunk 1 above)
Comment (liruipeng) [reply src/sstruct_mv/sstruct_matrix.c:1119]: @rfalgout There are a lot more we need to do to address 4D. GPU BoxLoops do not support 4D yet. Should we make the effort in another branch later? Thanks!

Comment (rfalgout) [reply src/sstruct_mv/sstruct_matrix.c:1119]: Okay, since it only affects GPUs.  Thanks!

**Comment by liruipeng** [comment]
@rfalgout I reverted some files that only had nonessential changes (white space/style changes). Most of the changed files in `sstruct_*` are related to `zypre_BoxLoopGetIndex`, while `sstruct_matrix.c` has most of the real changes in `sstruct`. The changes in `IJ` and `Parcsr` should not be too problematic to `recmat` branch. Hope this can help, to some extent.

**Comment by rfalgout** [comment]
Thanks @liruipeng !  Much appreciated!

**Comment by liruipeng** [comment]
@rfalgout @ulrikeyang @oseikuffuor1 Passed regression tests. This PR is ready to merge. Thanks!

**Review by rfalgout — APPROVED** [review]
Looks good Ruipeng!  For some reason I still can't view the diff for:

    src/sstruct_ls/HYPRE_sstruct_maxwell.c
    src/parcsr_mv/HYPRE_parcsr_vector.c
    src/parcsr_ls/HYPRE_parcsr_bicgstab.c
    src/krylov/pcg.c

I checked manually (with `git diff <previous commit>`, and there are not many changes in them, so this is weird.  I even looked to see if there are any strange hidden characters, but I don't see anything.  It may just be that github is confused for some reason.

Thanks!

**Review by oseikuffuor1 — APPROVED** [review]
Looks good. Thanks Ruipeng.

**Review by ulrikeyang — APPROVED** [review]
I can't say I looked at everything in detail, but I will approve this.

**Comment by liruipeng** [comment]
> Looks good Ruipeng! For some reason I still can't view the diff for:
> 
> ```
> src/sstruct_ls/HYPRE_sstruct_maxwell.c
> src/parcsr_mv/HYPRE_parcsr_vector.c
> src/parcsr_ls/HYPRE_parcsr_bicgstab.c
> src/krylov/pcg.c
> ```
> 
> I checked manually (with `git diff <previous commit>`, and there are not many changes in them, so this is weird. I even looked to see if there are any strange hidden characters, but I don't see anything. It may just be that github is confused for some reason.
> 
> Thanks!

Me neither. Github gets confused. I did the same thing you did. Thanks, @rfalgout

**Comment by ulrikeyang** [comment]
Yes, I had the same issue. It finally worked when I changed the settings to unified and did not check the hide white space button.

**Pull request merged**